winch/codegen/src/codegen/mod.rs

```diff
@@ -847,8 +847,29 @@ where
         size: OperandSize,
         sextend: Option<ExtendKind>,
     ) -> Result<()> {
-        let addr = self.emit_compute_heap_address(&arg, size)?;
-        if let Some(addr) = addr {
+        self.emit_wasm_load_full(arg, ty, size, sextend, false)
+    }
+
+    /// Emit a WebAssembly atomic load.
+    pub fn emit_wasm_load_atomic(
+        &mut self,
+        arg: &MemArg,
+        ty: WasmValType,
+        size: OperandSize,
+        sextend: Option<ExtendKind>,
+    ) -> Result<()> {
+        self.emit_wasm_load_full(arg, ty, size, sextend, true)
+    }
+
+    pub fn emit_wasm_load_full(
+        &mut self,
+        arg: &MemArg,
+        ty: WasmValType,
+        size: OperandSize,
+        sextend: Option<ExtendKind>,
+        atomic: bool,
```

Comment:
Rather than a `bool`, which can be error-prone, could you define a small enum `WasmLoadKind { Regular, Atomic }` or similar and match on it in this function?